pkg/apis/configuration/validation/validation.go

```diff
@@ -596,15 +596,140 @@ func validateRoute(route v1.Route, fieldPath *field.Path, upstreamNames sets.Str
+	if action.Pass != "" && action.Redirect != nil {
+		return append(allErrs, field.Required(fieldPath, "action must specify exactly one of `pass` or `redirect`"))
+	}
+
+	if action.Pass != "" {
+		allErrs = append(allErrs, validateReferencedUpstream(action.Pass, fieldPath.Child("pass"), upstreamNames)...)
+	}
+
+	if action.Redirect != nil {
+		allErrs = append(allErrs, validateActionRedirect(action.Redirect, fieldPath.Child("redirect"))...)
+	}
 
 	return allErrs
 }
 
+func validateActionRedirect(redirect *v1.ActionRedirect, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, validateRedirectURL(redirect.URL, fieldPath.Child("url"))...)
+
+	if redirect.Code != 0 {
+		allErrs = append(allErrs, validateRedirectStatusCode(redirect.Code, fieldPath.Child("code"))...)
+	}
+
+	return allErrs
+}
+
+// Returns a slice of vars enclosed in ${}. For example ${scheme} would return scheme.
+func captureVariables(s string) [][]string {
+	regex := regexp.MustCompile(`\$\{([^}]+)\}`)
+	return regex.FindAllStringSubmatch(s, -1)
+}
+
+// validRedirectVariableNames includes NGINX variables allowed to be used in redirects.
+var validRedirectVariableNames = map[string]bool{
+	"scheme":                 true,
+	"http_x_forwarded_proto": true,
+	"request_uri":            true,
+	"host":                   true,
+}
+
+func validateRedirectURL(redirectURL string, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if redirectURL == "" {
+		return append(allErrs, field.Required(fieldPath, "must specify a url"))
+	}
+
+	allErrs = append(allErrs, validateRedirectURLFmt(redirectURL, fieldPath)...)
+
+	if strings.Contains(redirectURL, "$") {
+		nginxVars := captureVariables(redirectURL)
+		if nginxVars != nil {
+			for _, v := range nginxVars {
+				nVar := v[1]
+				allErrs = append(allErrs, validateVariable(nVar, validRedirectVariableNames, fieldPath)...)
+			}
+		}
+	}
+
+	return allErrs
+}
+
+const redirectFmt string = `([^"\\]|\\.)*`
+const redirectFmtErrMsg string = `must have all '"' (double quotes) escaped and must not end with an unescaped '\' (backslash)`
+
+var redirectFmtRegexp = regexp.MustCompile("^" + redirectFmt + "$")
+
+func validateRedirectURLFmt(str string, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !redirectFmtRegexp.MatchString(str) {
+		msg := validation.RegexError(redirectFmtErrMsg, redirectFmt, "http://www.nginx.com", "$scheme://$host/green/", `\"http://www.nginx.com\"`)
+		return append(allErrs, field.Invalid(fieldPath, str, msg))
+	}
+
+	if strings.Contains(str, "${}") {
+		return append(allErrs, field.Invalid(fieldPath, str, "must not contain any empty variable ${}"))
+	}
+
+	if strings.HasSuffix(str, "$") {
+		return append(allErrs, field.Invalid(fieldPath, str, "must not end with $"))
+	}
+
+	for i, c := range str {
+		if c == '$' && i+1 < len(str) {
```

Comment:
I was thinking, `&& i+1 < len(str)` is not really needed because in line 683 we return if the last character is `$` and that's the only possible panic, when the `$` is at the end of the string. So unless I am missing something, up to you to remove it.